### ISSUE TITLE
Unreal Marketplace Testing Failure for Android for v1.3.0

### ISSUE THREAD

**Issue by shehzan10** [CesiumGS/cesium-unreal#472]
For v1.3.0, we added Android to the whitelist platform in the uplugin file. This is causing them to run into C++17 compilation issues similar to https://github.com/CesiumGS/cesium-unreal/issues/190#issuecomment-831522727.

[Cesium Errors.txt](https://github.com/CesiumGS/cesium-unreal/files/6579898/Cesium.Errors.txt)

If we don't need Android support through marketplace for this release, they are recommending that we submit the plugin by removing `Android` from the uplugin Supported and Target platforms, and instead move it to `BlacklistPlatforms` for this release.

Regarding the errors: As mentioned in the issue comment above, I did see them when trying to build the plugin for Android at one point, but they essentially disappeared one day. I'm still not sure why they don't show up on CI. Those errors also popped up when I was trying to compile Android with UE5 EA over the weekend.

CC @CesiumGS/unreal

**Comment by shehzan10** [comment]
For the build submitted for v1.3.1:

```
While attempting to process your file update, we encountered the following errors: 

/CesiumForUnreal/Binaries/Mac/UE4Editor-CesiumRuntime.dylib for architecture x86_64
4706
[00:01:42]
clang: error: linker command failed with exit code 1 (use -v to see invocation)

Notes:
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation).
Possible solution: A missing vtable usually means the first non-inline virtual member function has no definition.
```

@baothientran Can you look into this?

**Comment by baothientran** [comment]
sure I will look into it right away

**Comment by baothientran** [comment]
@shehzan10 I can't reproduce the error even doing it from scratch. Can we ask them what steps they follow to produce that error? Also another thing we can suggest is to make sure security enable for CesiumRuntime.dylib and CesiumEditor.dylib. But I dout it's an issue

**Comment by shehzan10** [comment]
The Mac error from https://github.com/CesiumGS/cesium-unreal/issues/472#issuecomment-853104369 seemed to be a false warning. We need need to figure out the Android compilation issue.

**Comment by drakelzh** [comment]
> For v1.3.0, we added Android to the whitelist platform in the uplugin file. This is causing them to run into C++17 compilation issues similar to [#190 (comment)](https://github.com/CesiumGS/cesium-unreal/issues/190#issuecomment-831522727).
> 
> [Cesium Errors.txt](https://github.com/CesiumGS/cesium-unreal/files/6579898/Cesium.Errors.txt)
> 
> If we don't need Android support through marketplace for this release, they are recommending that we submit the plugin by removing `Android` from the uplugin Supported and Target platforms, and instead move it to `BlacklistPlatforms` for this release.
> 
> Regarding the errors: As mentioned in the issue comment above, I did see them when trying to build the plugin for Android at one point, but they essentially disappeared one day. I'm still not sure why they don't show up on CI. Those errors also popped up when I was trying to compile Android with UE5 EA over the weekend.
> 
> CC @CesiumGS/unreal

After many attempts of mine, the C++17 problem of Android packaging can be corrected by selecting the fourth item in Project Settings-Project-Packaging-Compile (the third item is selected by default), and selecting the fourth item seems to be mandatory Use a local compiler, but you need to add in build.cs:
// Tell Unreal Engine to use C++17
         CppStandard = CppStandardVersion.Cpp17;
I have successfully packaged and installed it successfully on Android, but it crashed immediately after startup. I found that the real problem was the error prompted after the following operation: In Unreal, you can change the render settings to emulate your Android settings. When I do this, it crashes for me and gives me this error: Assertion failed: Stream.VertexBuffer->IsInitialized() [File:D:/Build/++UE4/Sync/Engine/Source/Runtime/RenderCore/Private/ VertexFactory.cpp] [Line: 210]
Vertex buffer was not initialized! Stream 3, Stride 4, Name ColorOnly Mesh Vertices.

**Comment by kring** [comment]
@drakelzh I believe we fixed that Vertex buffer issue quite awhile ago. Are you using Cesium for Unreal v1.3.1 or a recent version of main?

**Comment by drakelzh** [comment]
@kring Yes, you are correct. I’m sorry. I forgot to check the latest plugin version. The problem no longer exists. Thank you for your team’s efforts.

**Comment by xuelongmu** [comment]
@shehzan10 I am attempting to locally package the plugin for Android on the current `main` branch. However I am running into the mikktspace dependency error that was discussed in #444

Beforehand, I compiled the extern directory with VS 2017, and then ran the `Windows + Android` example given in the [packaging documentation](https://github.com/CesiumGS/cesium-unreal/blob/main/Documentation/packaging-guide.md)

Got the following error:
```
  Reflection code generated for UE4Game in 2.5432914 seconds
  Compiling Native 32-bit code with NDK API 'android-21'
  Compiling Native 64-bit code with NDK API 'android-21'
  Writing manifest to C:\Users\Xuelong\Dev\CesiumDevTest\Packages\HostProject\Saved\Manifest-UE4Game-Android-Development.xml
  Building UE4Game...
  Building 6 actions with 12 processes...
    [1/6] PCH.CesiumRuntime.h [armv7]
    [2/6] PCH.CesiumRuntime.h [arm64]
    [3/6] Module.CesiumRuntime.gen.cpp [armv7]
    [4/6] Module.CesiumRuntime.gen.cpp [arm64]
    [5/6] Module.CesiumRuntime.cpp [armv7]
    In file included from C:/Users/Xuelong/Dev/CesiumDevTest/Packages/HostProject/Plugins/CesiumForUnreal/Intermediate/Build/Android/UE4/Development/CesiumRuntime/Module.CesiumRuntime.cpp:12:
    C:/Users/Xuelong/Dev/CesiumDevTest/Packages/HostProject/Plugins/CesiumForUnreal/Source/CesiumRuntime/Private/CesiumGltfComponent.cpp(37,10): fatal error: 'mikktspace.h' file not found
    #include "mikktspace.h"
             ^~~~~~~~~~~~~~
    1 error generated.
    [6/6] Module.CesiumRuntime.cpp [arm64]
    In file included from C:/Users/Xuelong/Dev/CesiumDevTest/Packages/HostProject/Plugins/CesiumForUnreal/Intermediate/Build/Android/UE4/Development/CesiumRuntime/Module.CesiumRuntime.cpp:12:
    C:/Users/Xuelong/Dev/CesiumDevTest/Packages/HostProject/Plugins/CesiumForUnreal/Source/CesiumRuntime/Private/CesiumGltfComponent.cpp(37,10): fatal error: 'mikktspace.h' file not found
    #include "mikktspace.h"
             ^~~~~~~~~~~~~~
    1 error generated.
Took 19.8765484s to run UnrealBuildTool.exe, ExitCode=6
UnrealBuildTool failed. See log for more details. (C:\Users\Xuelong\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_4.26\UBT-UE4Game-Android-Development.txt)
AutomationTool exiting with ExitCode=6 (6)
BUILD FAILED
```

**Comment by shehzan10** [comment]
@xuelongmu If you were using an existing repo, you may need to do a `git submodule update --init --recursive` from the cesium-unreal repo. `MikkTSpace` was added as a submodule to Unreal, not native.

**Comment by xuelongmu** [comment]
@shehzan10 I verified by running `git submodule update --init --recursive` command from the cesium-unreal repo (which didn't produce any log messages), checking that the MikkTSpace folder existed, and ran the build again with the same error.

**Comment by kring** [comment]
> Beforehand, I compiled the extern directory with VS 2017

Just to check the simple stuff... You need to _install_ the extern directory, not just compile it.

**Comment by kring** [comment]
As far as we can tell, the Android build is working well. I'm going to close this, and we'll open a new one if Epic is still unable to package our plugin next time we publish.

**Comment by kring** [comment]
Epic is still having the same problem on Android. But so far no luck getting a complete log from them. Some things to try:

* Put the package we provide inside a new project, and then try to package it from there using the Editor UI. Rather than putting it in a separate directory and packaging from the command line.
* Are there any dodgy project settings that could cause problems when we do the above?
* Maybe the path is too long? Imitate Epic's paths as shown in the log file.
* What if Epic is unzipping our plugin package with the built-in Windows unzip that marks all the files as unsafe because they're from the internet?
* What if Epic is marking the files in our ZIP read-only?

**Comment by kring** [comment]
I did the following:

* Created a new project in `C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\HostProject` to match the paths in the Epic log, except I don't have a D drive so I put it on C.
* Extracted our CI-produced ZIP file to the Plugins subdirectory of that. I used the Windows unzip to extract it.
* Opened the project in the Editor, went to Edit -> Plugins, found the Cesium plugin, and clicked the Package button.

I did this on Windows, and the plugin built successfully for Windows, Linux, and Android. That torpedos a bunch of my ideas above.

**Comment by kring** [comment]
Making the plugin files readonly after extracting them does cause the build to fail, but in a different and fairly predictable way:

```
UATHelper: Package Plugin Task (Windows): Running AutomationTool...
UATHelper: Package Plugin Task (Windows): Parsing command line: BuildPlugin -Plugin=C:/Build/++Portal/Sync/LocalBuilds/PluginTemp/HostProject/Plugins/CesiumForUnreal/CesiumForUnreal.uplugin -Package=C:/Build/++Portal/Sync/LocalBuilds/PluginTemp/package/CesiumForUnreal -CreateSubFolder
UATHelper: Package Plugin Task (Windows): Copying 1242 file(s) using max 64 thread(s)
UATHelper: Package Plugin Task (Windows): Reading plugin from C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Plugins\CesiumForUnreal\CesiumForUnreal.uplugin...
UATHelper: Package Plugin Task (Windows): Building plugin for host platforms: Win64
UATHelper: Package Plugin Task (Windows): Running: C:\Program Files\Epic Games\UE_4.26\Engine\Binaries\DotNET\UnrealBuildTool.exe UE4Editor Win64 Development -Project=C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\HostProject.uproject -plugin=C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\
Plugins\CesiumForUnreal\CesiumForUnreal.uplugin -iwyu -noubtmakefiles -manifest=C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Saved\Manifest-UE4Editor-Win64-Development.xml -nohotreload -log="C:\Users\kring\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_4.26\UBT-UE4Editor-Win64-Development.
txt"
UATHelper: Package Plugin Task (Windows):   Parsing headers for UE4Editor
UATHelper: Package Plugin Task (Windows):     Running UnrealHeaderTool "C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\HostProject.uproject" "C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Intermediate\Build\Win64\UE4Editor\Development\UE4Editor.uhtmanifest" -LogCmds="loginit warning, loge
xit warning, logdatabase error" -Unattended -WarningsAsErrors -abslog="C:\Users\kring\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_4.26\UHT-UE4Editor-Win64-Development.txt" -installed
UATHelper: Package Plugin Task (Windows):   Reflection code generated for UE4Editor in 8.242515 seconds
UATHelper: Package Plugin Task (Windows):   Writing manifest to C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Saved\Manifest-UE4Editor-Win64-Development.xml
UATHelper: Package Plugin Task (Windows):   Building UE4Editor...
UATHelper: Package Plugin Task (Windows):   Using Visual Studio 2019 14.29.30038 toolchain (C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30037) and Windows 10.0.19041.0 SDK (C:\Program Files (x86)\Windows Kits\10).
UATHelper: Package Plugin Task (Windows):   Building 13 actions with 12 processes...
UATHelper: Package Plugin Task (Windows):     [1/13] Default.rc2
UATHelper: Package Plugin Task (Windows):     [2/13] Default.rc2
UATHelper: Package Plugin Task (Windows):     [3/13] PCH.CesiumRuntime.cpp
UATHelper: Package Plugin Task (Windows):     [4/13] PCH.CesiumEditor.cpp
UATHelper: Package Plugin Task (Windows):     [5/13] Module.CesiumEditor.gen.cpp
UATHelper: Package Plugin Task (Windows):     [6/13] Module.CesiumRuntime.gen.cpp
UATHelper: Package Plugin Task (Windows):     [7/13] Module.CesiumRuntime.cpp
UATHelper: Package Plugin Task (Windows):     [8/13] UE4Editor-CesiumRuntime.lib
UATHelper: Package Plugin Task (Windows):        Creating library C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Plugins\CesiumForUnreal\Intermediate\Build\Win64\UE4Editor\Development\CesiumRuntime\UE4Editor-CesiumRuntime.lib and object C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Plugin
s\CesiumForUnreal\Intermediate\Build\Win64\UE4Editor\Development\CesiumRuntime\UE4Editor-CesiumRuntime.exp
UATHelper: Package Plugin Task (Windows):     [9/13] UE4Editor-CesiumRuntime.dll
UATHelper: Package Plugin Task (Windows):     LINK : fatal error LNK1104: cannot open file 'C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Plugins\CesiumForUnreal\Binaries\Win64\UE4Editor-CesiumRuntime.dll'
UATHelper: Package Plugin Task (Windows):     [10/13] Module.CesiumEditor.cpp
UATHelper: Package Plugin Task (Windows):     [11/13] UE4Editor-CesiumEditor.lib
UATHelper: Package Plugin Task (Windows):        Creating library C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Plugins\CesiumForUnreal\Intermediate\Build\Win64\UE4Editor\Development\CesiumEditor\UE4Editor-CesiumEditor.lib and object C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Plugins\
CesiumForUnreal\Intermediate\Build\Win64\UE4Editor\Development\CesiumEditor\UE4Editor-CesiumEditor.exp
UATHelper: Package Plugin Task (Windows):     [12/13] UE4Editor-CesiumEditor.dll
UATHelper: Package Plugin Task (Windows):     LINK : fatal error LNK1104: cannot open file 'C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Plugins\CesiumForUnreal\Binaries\Win64\UE4Editor-CesiumEditor.dll'
UATHelper: Package Plugin Task (Windows): Took 53.875737s to run UnrealBuildTool.exe, ExitCode=6
UATHelper: Package Plugin Task (Windows): UnrealBuildTool failed. See log for more details. (C:\Users\kring\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_4.26\UBT-UE4Editor-Win64-Development.txt)
UATHelper: Package Plugin Task (Windows): AutomationTool exiting with ExitCode=6 (6)
UATHelper: Package Plugin Task (Windows): BUILD FAILED
```

**Comment by xuelongmu** [comment]
> Created a new project in C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\HostProject to match the paths in the Epic log,

Do you happen to have long file paths enabled on your machine? Maybe you could try disabling it https://www.howtogeek.com/266621/how-to-make-windows-10-accept-file-paths-over-260-characters/

**Comment by kring** [comment]
Oh that's a really good point, thanks @xuelongmu! I do have it enabled. I'll retry that test with it disabled.

**Comment by kring** [comment]
Well that was a great thought, but I changed the registry setting, rebooted, and re-ran the test from the beginning, and the plugin still builds successfully on all platforms.

**Comment by kring** [comment]
Epic finally sent me a log. The main difference that jumps out at me in the Android build is that we have a slightly newer version of the Android NDK.

Ours:

```
PLATFORM_ANDROID_NDK_VERSION = 210500
NDK toolchain: r21e, NDK version: 21, GccVersion: 4.9, ClangVersion: 9.0.9
```

Theirs:

```
PLATFORM_ANDROID_NDK_VERSION = 210200
NDK toolchain: r21b, NDK version: 21, GccVersion: 4.9, ClangVersion: 9.0.8
```

Another (less likely) thing is that their build is done with XGE / Incredibuild, whereas ours is not.

I'll see if I can reproduce their build failure with this new information.

**Comment by kring** [comment]
I downgrade to 21b in the https://github.com/CesiumGS/cesium-unreal/tree/android-ndk-21b branch, and now it reports exactly the same NDK information as reported in Epic's log. But our build still completes successfully.

**Comment by kring** [comment]
When the build fails, Epic's log reports:
```
Error executing ...\Android\-23\ndk\21.0.6113669\toolchains\llvm\prebuilt\windows-x86_64\bin\clang++.exe (tool returned code: 1)
```

That `21.0.6113669` is interesting because it doesn't correspond to version 21b of the Android NDK (that would be `21.1.6352462`), it corresponds to the original version 21 (you might call it 21a, but Google doesn't). So despite the log saying that it's using version 21b of the Android NDK, Epic's XGE-based build is apparently actually use 21. Ok, guess I'll try that one. Based on the NDK changelog, though, there's no indication that C++17 support would be available in 21b but not in 21. So the more likely explanation is that the XGE-based build is misconfigured in other ways, not _just_ using an unexpected version of the NDK.

**Comment by kring** [comment]
I installed Incredibuild / XGE on my system, and now the UE plugin package process uses it. But still it completes successfully. Two caveats, though: I'm still using NDK 21e locally rather than NDK 21, and I don't have an Incredibuild license so I'm running in "Standalone Mode". I don't _think_ either of those should matter, but who knows.

**Comment by kring** [comment]
I finally figured out the cause of the build failures. Unreal Engine gained support for C++17 on Android in version 4.26.2. Prior versions, including 4.26.0, and 4.26.1, only supported C++14 on Android. This was a limitation in the Unreal Build Tool, not anything deeper than that. But unfortunately, Epic's Marketplace build process uses 4.26.0, so we can't use C++17 in a Marketplace plugin that supports Android. Options:

1. Convince Epic to use 4.26.2 in their Marketplace plugin build process. I'm not going to hold my breath.
2. For anyone that wants to use our plugin on Android, make them install it from GitHub rather than from the Marketplace. Not ideal.
3. Change Cesium for Unreal and cesium-native so that it can compiler for C++14. At least the non-Editor bits, because the Editor doesn't run on Android anyway. I started this in https://github.com/CesiumGS/cesium-native/pull/280.
